plugins/metrics/src/main/java/org/apache/cloudstack/metrics/MetricsServiceImpl.java

```diff
@@ -134,6 +155,166 @@ private void updateHostMetrics(final Metrics metrics, final HostJoinVO host) {
+        if (CollectionUtils.isNotEmpty(ids)) {
+            sc.setParameters("idIN", ids.toArray());
+        }
+        if (StringUtils.isNotBlank(name)) {
+            sc.setParameters("displayName", "%" + name + "%");
+        }
+        if (StringUtils.isNotBlank(keyword)) {
+            SearchCriteria<UserVmVO> ssc = userVmDao.createSearchCriteria();
+            ssc.addOr("displayName", SearchCriteria.Op.LIKE, "%" + keyword + "%");
+            ssc.addOr("state", SearchCriteria.Op.EQ, keyword);
+            sc.addAnd("displayName", SearchCriteria.Op.SC, ssc);
+        }
+
+        return userVmDao.searchAndCount(sc, searchFilter);
+    }
+
+    /**
+     * Searches stats for a list of VMs, based on date filtering parameters.
+     *
+     * @param cmd the {@link ListVMsUsageHistoryCmd} specifying the filtering parameters.
+     * @param userVmList the list of VMs for which stats should be searched.
+     * @return the key-value map in which keys are VM IDs and values are lists of VM stats.
+     */
+    protected Map<Long,List<VmStatsVO>> searchForVmMetricsStatsInternal(ListVMsUsageHistoryCmd cmd, List<UserVmVO> userVmList) {
+        Map<Long,List<VmStatsVO>> vmStatsVOList = new HashMap<Long,List<VmStatsVO>>();
+        Date startDate = cmd.getStartDate();
+        Date endDate = cmd.getEndDate();
+
+        validateDateParams(startDate, endDate);
+
+        for (UserVmVO userVmVO : userVmList) {
+            Long vmId = userVmVO.getId();
+            vmStatsVOList.put(vmId, findVmStatsAccordingToDateParams(vmId, startDate, endDate));
+        }
+
+        return vmStatsVOList;
+    }
+
+    /**
+     * Checks if {@code startDate} is after {@code endDate} (when both are not null)
+     * and throws an {@link InvalidParameterValueException} if so.
+     *
+     * @param startDate the start date to be validated.
+     * @param endDate the end date to be validated.
+     */
+    protected void validateDateParams(Date startDate, Date endDate) {
+        if ((startDate != null && endDate != null) && (startDate.after(endDate))){
+            throw new InvalidParameterValueException("startDate cannot be after endDate.");
+        }
+    }
+
+    /**
+     * Finds stats for a specific VM based on date parameters.
+     *
+     * @param vmId the specific VM.
+     * @param startDate the start date to filtering.
+     * @param endDate the end date to filtering.
+     * @return the list of stats for the specified VM.
+     */
+    protected List<VmStatsVO> findVmStatsAccordingToDateParams(Long vmId, Date startDate, Date endDate){
+        if (startDate != null && endDate != null) {
+            return vmStatsDao.findByVmIdAndTimestampBetween(vmId, startDate, endDate);
+        } else if (startDate != null) {
+            return vmStatsDao.findByVmIdAndTimestampGreaterThanEqual(vmId, startDate);
+        } else if (endDate != null) {
+            return vmStatsDao.findByVmIdAndTimestampLessThanEqual(vmId, endDate);
+        }
```

Comment:
If the block returns something, then we don't need the else statement. We could remove them:

```suggestion
        if (startDate != null && endDate != null) {
            return vmStatsDao.findByVmIdAndTimestampBetween(vmId, startDate, endDate);
        }

        if (startDate != null) {
            return vmStatsDao.findByVmIdAndTimestampGreaterThanEqual(vmId, startDate);
        }

        if (endDate != null) {
            return vmStatsDao.findByVmIdAndTimestampLessThanEqual(vmId, endDate);
        }
```